galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt

```diff
@@ -2610,16 +2610,43 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
       val id = if (isRecordFetched(recno)) idField.getInt(recno) else -1
 
       for (i in indices!!.indices) {
-        checkUniqueIndex(i, recno, id)
+        checkUniqueIndex(i, recno, id!!)
       }
     }
   }
 
   /*
    * Checks unique index constraints
    */
-  protected fun checkUniqueIndex(idx: Int, recno: Int, id: Int?) {
-    TODO()
+  protected open fun checkUniqueIndex(index: Int, recno: Int, id: Int) {
+    val condition = mutableListOf<Op<Boolean>>()
+
+    for (field in fields) {
+      val column  = if (field.isNull(recno) || !field.hasIndex(index)) {
+        null
+      } else {
+        @Suppress("UNCHECKED_CAST")
+        field.lookupColumn(0) as? Column<Any>
+      }
+      if (column != null) {
+        condition.add(Op.build { column eq field.getSql(recno)!! })
+      }
+    }
+    if (condition.isNotEmpty()) {
+      val query = tables!![0].slice(idColumn).select{ condition.compoundAnd() }
+
+      transaction {
+        if (query.execute(this)!!.next()) {
+          if (query.first()[idColumn] != id) {
+            form.setActiveBlock(this@VBlock)
+            activeRecord = recno
+            gotoFirstField()
+            throw VExecFailedException(MessageCode.getMessage("VIS-00014", arrayOf<Any>(indices!![index])))
+          }
+          assert(!query.execute(this)!!.next()) { "too many rows" }
+        }
```

Comment:
use single() method, and catch `NoSuchElementException` and `IllegalArgumentException` seperately with corresponding error message